core/src/test/scala/pureconfig/PureconfSuite.scala

```diff
@@ -413,6 +413,40 @@ class PureconfSuite extends FlatSpec with Matchers with OptionValues with TryVal
     loadConfig[ConfWithListOfFoo](conf) shouldBe Success(expected)
   }
 
+  it should "be able to load a set of Ints from the system properties" in {
+    val conf = ConfigFactory.parseString("""
+    pure.conf: {
+      intSet.0: 1
+      intSet.1: 2
+    }""")
+
+    val expected = Set(1, 2)
+    loadConfig[Set[Int]](conf, "pure.conf.intSet") shouldBe Success(expected)
+  }
+
+  it should "be able to load a list of Ints from the system properties in correct order" in {
+    val conf = ConfigFactory.parseString("""
+    pure.conf: {
+      intList.2: 1
+      intList.0: 2
+      intList.1: 3
+    }""")
+
+    val expected = List(2, 3, 1)
+    loadConfig[List[Int]](conf, "pure.conf.intList") shouldBe Success(expected)
+  }
+
+  it should "be able to load a list of Ints from the system properties in correct order when one element is missing" in {
+    val conf = ConfigFactory.parseString("""
+    pure.conf: {
+      intList.2: 3
+      intList.0: 1
+    }""")
+
+    val expected = List(1, 3)
```

Comment:
I think the behavior is smart in the sense that what the number is is not important because the number is not used as index in the structure. For what I can see, it could also be not a number: given a type `T`, if it has an `Ordering[T]` then you can use it as index because you just need it to sort the key-values.